src/lib/pbkdf/argon2/argon2.cpp

```diff
@@ -296,6 +297,10 @@ void process_block(secure_vector<uint64_t>& B,
    }
 
    while(index < segments) {
+      if((index & 63) == 0 && stop_token.has_value() && stop_token->stop_requested()) {
+         throw Botan::Invalid_State("Cancelled");
```

Comment:
I agree that `Invalid_State` is probably the closest exception type for this. But I would argue that it is worthwhile to introduce a new exception type `Operation_Canceled` that inherits from `Invalid_State` so that applications can explicitly distinguish cancellations.

Reply:
My reasoning for using Invalid_State was that the caller can easily distinguish the cancelled state, because the caller owns the stop_source.

In my application, I currently call it like this:

```
try {
            passwordHash->derive_key(..., stop_source.get_token());
        } catch (Botan::Exception& e) {
            if (stop_source.stop_requested()) {
                 // handle cancellation
            } else {
                 // handle other error
            }
        }
```

Adding Operation_Cancelled is nice, that simplifies the catch to simply catching Botan:: Operation_Cancelled.

I'll add it in the next commit.

Reply:
I see I didn't add that logic to the unit test, my bad.. Anyway, will be fixed after adding the new exception.

Reply:
Operation_Canceled has been added in commit 0e63412

Reply:
In ffi.cpp I mapped OperationCanceled to BOTAN_FFI_ERROR_INVALID_OBJECT_STATE, as the stop_token is not exposed to FFI, so this error cannot occur inside ffi (currently).

If preferred I can add a BOTAN_FFI_ERROR_OPERATION_CANCELED there but that might confuse consumers of the ffi interface.